pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexType.java

```diff
@@ -0,0 +1,171 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.spi.index;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import java.io.IOException;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.apache.pinot.segment.spi.ColumnMetadata;
+import org.apache.pinot.segment.spi.creator.IndexCreationContext;
+import org.apache.pinot.segment.spi.index.column.ColumnIndexContainer;
+import org.apache.pinot.segment.spi.store.SegmentDirectory;
+import org.apache.pinot.spi.config.table.FieldConfig;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.utils.JsonUtils;
+
+
+/**
+ * TODO: implement mutable indexes.
+ * @param <C> the class that represents how this object is configured.
+ * @param <IR> the {@link IndexReader} subclass that should be used to read indexes of this type.
+ * @param <IC> the {@link IndexCreator} subclass that should be used to create indexes of this type.
+ */
+public interface IndexType<C, IR extends IndexReader, IC extends IndexCreator> {
+
+  /**
+   * The unique id that identifies this index type.
+   * In case there is more than one implementation for a given index, then all should share the same id in order to be
+   * correctly registered in the {@link IndexService}.
+   * This is also the value being used as the default toString implementation and the one used as keys when config is
+   * specified.
+   *
+   * <p>Therefore the returned value for each index should be constant across different Pinot versions.</p>
+   */
+  String getId();
+
+  /**
+   * Returns an internal name used in some parts of the code (mainly in format v1 and metadata) that is persisted on
+   * disk.
+   *
+   * <p>Therefore the returned value for each index should be constant across different Pinot versions.</p>
+   */
+  String getIndexName();
+
+  default Class<C> getIndexConfigClass() {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * The default config when it is not explicitly defined by the user.
+   *
+   * Can return null if the index should be disabled by default.
+   */
+  @Nullable
+  default C getDefaultConfig() {
+    return null;
+  }
+
+  /**
+   * This method is called to transform from a JSON node to a config object.
+   *
+   * This is usually used to deserialize {@link FieldConfig#getIndexes() fieldConfigLists.indexes.(indexId)}.
+   * @throws IOException
+   */
+  @Nullable
```

Comment:
Is this nullable?

Reply:
Yes, this is nullable. In fact you have to return null if the config indicates that the index should be disabled.